### PR TITLE
[MXNET-394] concat of CSR NDArrays on first dimension

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
concat of CSR NDArrays on first dimension

cc @eric-haibin-lin 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
CI is failing

**Comment by eric-haibin-lin** [comment]
@haojin2 could you help review this? The stack op is no longer needed for csr

File: src/operator/nn/concat-inl.h [hunk 1]
```diff
@@ -154,6 +154,82 @@ void ConcatGradCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
...
+    out_indptr[i+1+indptr_offset] = in_indptr[i+1] + idx_offset;
+    for (nnvm::dim_t j = in_indptr[i]; j < in_indptr[i+1]; ++j) {
+      KERNEL_ASSIGN(out_idx[j+idx_offset], req, in_idx[j]);
+      KERNEL_ASSIGN(out_data[j+idx_offset], req, in_data[j]);
+    }
+  }
+};
+
+
+
```
Comment (haojin2) [inline src/operator/nn/concat-inl.h]: nit: too many blank lines here

File: src/operator/nn/concat-inl.h [hunk 2]
```diff
@@ -154,6 +154,82 @@ void ConcatGradCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
   });
 }
 
+
```
Comment (haojin2) [inline src/operator/nn/concat-inl.h]: nit: too many blank lines here

File: src/operator/nn/concat.cc [hunk 3]
```diff
@@ -105,27 +105,41 @@ static bool ConcatType(const nnvm::NodeAttrs& attrs,
   return true;
 }
 
+
```
Comment (haojin2) [inline src/operator/nn/concat.cc]: nit: too many blank lines here

File: src/operator/nn/concat.cc [hunk 4]
```diff
@@ -105,27 +105,41 @@ static bool ConcatType(const nnvm::NodeAttrs& attrs,
...
+    dispatched = storage_type_assign(&out_stype, kDefaultStorage,
+                                     dispatch_mode, DispatchMode::kFCompute);
+  }
+  if (!dispatched) {
+    dispatched = dispatch_fallback(out_attrs, dispatch_mode);
+  }
+  return dispatched;
 }
 
+
```
Comment (haojin2) [inline src/operator/nn/concat.cc]: nit: too many blank lines here

File: src/operator/nn/concat.cc [hunk 5]
```diff
@@ -146,7 +160,7 @@ inline static bool BackwardConcatStorageType(const nnvm::NodeAttrs& attrs,
                              dispatch_mode, wanted_mode);
 }
 
-#if MXNET_USE_MKLDNN == 1
+
```
Comment (haojin2) [inline src/operator/nn/concat.cc]: nit: too many blank lines here

File: tests/python/unittest/test_sparse_ndarray.py [hunk 6]
```diff
@@ -155,6 +155,22 @@ def check_slice_nd_csr_fallback(shape):
...
+    zero_nds = []
+    ncols = rnd.randint(2, 10)
+    for i in range(3):
+        shape = (rnd.randint(2, 10), ncols)
+        A, _ = rand_sparse_ndarray(shape, 'csr')
+        nds.append(A)
+        zero_nds.append(mx.nd.zeros(shape).tostype('csr'))
+    check_concat(nds)
+    check_concat(zero_nds)
+
```
Comment (haojin2) [inline tests/python/unittest/test_sparse_ndarray.py]: nit: one more blank line here

File: src/operator/nn/concat-inl.h [hunk 7]
```diff
@@ -154,6 +154,82 @@ void ConcatGradCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
...
+
+
+template<typename xpu>
+void ConcatRspImpl(const nnvm::NodeAttrs& attrs,
+                   const OpContext& ctx,
+                   const std::vector<NDArray>& inputs,
+                   const std::vector<OpReqType>& req,
+                   const std::vector<NDArray>& outputs) {
+  if (req[0] == kNullOp) return;
```
Comment (haojin2) [inline src/operator/nn/concat-inl.h]: Better do this after all the checks below?

File: src/operator/nn/concat-inl.h [hunk 8]
```diff
@@ -154,6 +154,82 @@ void ConcatGradCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
   });
 }
 
+
+struct concat_csr_first_dim {
+  // i is the i-th row of the output ndarray
```
Comment (haojin2) [inline src/operator/nn/concat-inl.h]: Better add more comments for all arguments?

**Comment by ZiyueHuang** [comment]
Anything remaining to address? @eric-haibin-lin @haojin2

**Comment by haojin2** [comment]
LGTM, let's wait for @eric-haibin-lin to take a final look

File: src/operator/nn/concat.cc [hunk 9]
```diff
@@ -261,9 +278,7 @@ Example::
 .set_attr<nnvm::FInferType>("FInferType", ConcatType)
 .set_attr<FInferStorageType>("FInferStorageType", ConcatForwardInferStorageType)
 .set_attr<FCompute>("FCompute<cpu>", ConcatCompute<cpu>)
-#if MXNET_USE_MKLDNN == 1
```
Comment (eric-haibin-lin) [inline src/operator/nn/concat.cc]: pls update doc

Comment (ZiyueHuang) [reply src/operator/nn/concat.cc]: Have updated several docs and add sparse op alias.

File: src/operator/nn/concat-inl.h [hunk 10]
```diff
@@ -154,6 +154,92 @@ void ConcatGradCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
...
+  using namespace mshadow;
+  using namespace mxnet_op;
+  using namespace csr;
+  const ConcatParam& param = nnvm::get<ConcatParam>(attrs.parsed);
+  int num_args = param.num_args;
+  int concat_dim = param.dim;
+  CHECK_EQ(inputs.size(), num_args);
+  CHECK_EQ(outputs.size(), 1);
+  int axis = CheckAxis(concat_dim, inputs[0].shape().ndim());
+  CHECK_EQ(axis, 0);
```
Comment (eric-haibin-lin) [inline src/operator/nn/concat-inl.h]: Pls add detailed err msg

File: src/operator/nn/concat-inl.h [hunk 11]
```diff
@@ -154,6 +154,92 @@ void ConcatGradCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
...
+  const NDArray& out = outputs[0];
+  const nnvm::dim_t num_rows = out.shape()[0];
+  out.CheckAndAllocAuxData(kIndPtr, Shape1(num_rows+1));
+
+  MSHADOW_IDX_TYPE_SWITCH(inputs[0].aux_type(kIndPtr), RType, {
+    MSHADOW_IDX_TYPE_SWITCH(inputs[0].aux_type(kIdx), IType, {
+      MSHADOW_TYPE_SWITCH(inputs[0].dtype(), DType, {
+        RType* out_indptr = out.aux_data(kIndPtr).dptr<RType>();
+        if (nnz == 0) {
+          out.set_aux_shape(kIdx, Shape1(0));
```
Comment (eric-haibin-lin) [inline src/operator/nn/concat-inl.h]: Better to directly call `FillZerosCsrImpl`?

File: src/operator/nn/concat-inl.h [hunk 12]
```diff
@@ -154,6 +154,92 @@ void ConcatGradCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
...
+          return;
+        }
+        out.CheckAndAllocAuxData(kIdx, Shape1(nnz));
+        out.CheckAndAllocData(Shape1(nnz));
+        IType* out_idx = out.aux_data(kIdx).dptr<IType>();
+        DType* out_data = out.data().dptr<DType>();
+        nnvm::dim_t indptr_offset = 0;
+        nnvm::dim_t idx_offset = 0;
+        for (const auto& in : inputs) {
+          RType* in_indptr = in.aux_data(kIndPtr).dptr<RType>();
```
Comment (eric-haibin-lin) [inline src/operator/nn/concat-inl.h]: nit: const

File: src/operator/nn/concat.cc [hunk 13]
```diff
@@ -112,18 +112,30 @@ inline static bool ConcatForwardInferStorageType(const nnvm::NodeAttrs& attrs,
...
+    dispatched = storage_type_assign(&out_stype, kCSRStorage,
+                                     dispatch_mode, DispatchMode::kFComputeEx);
+  }
+#if MXNET_USE_MKLDNN == 1
+  if (!dispatched && dev_mask == mshadow::cpu::kDevMask
       && common::ContainsOnlyStorage(*in_attrs, kDefaultStorage)
-      && param.dim > 0)
-    wanted_mode = DispatchMode::kFComputeEx;
-  else
+      && param.dim > 0) {
```
Comment (eric-haibin-lin) [inline src/operator/nn/concat.cc]: @zheng-da @pengzhao-intel is mkldnn concat only limited to dim > 0? Should be trivial to support negative dims I think

File: src/operator/nn/concat-inl.h [hunk 14]
```diff
@@ -154,6 +154,92 @@ void ConcatGradCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
...
+    out_indptr[i+1+indptr_offset] = in_indptr[i+1] + idx_offset;
+    for (nnvm::dim_t j = in_indptr[i]; j < in_indptr[i+1]; ++j) {
+      KERNEL_ASSIGN(out_idx[j+idx_offset], req, in_idx[j]);
+      KERNEL_ASSIGN(out_data[j+idx_offset], req, in_data[j]);
+    }
+  }
+};
+
+template<typename xpu>
+void ConcatRspImpl(const nnvm::NodeAttrs& attrs,
```
Comment (eric-haibin-lin) [inline src/operator/nn/concat-inl.h]: Is this CSR Impl or RSP impl??

File: src/operator/nn/concat.cc [hunk 15]
```diff
@@ -156,17 +167,23 @@ static void ConcatComputeExCPU(const nnvm::NodeAttrs& attrs,
   CHECK_EQ(outputs.size(), 1U);
   CHECK_EQ(req.size(), 1U);
   if (req[0] == kNullOp) return;
-  // MKLDNN support 2D and 4D concat
-  if ((inputs[0].shape().ndim() == 2 || inputs[0].shape().ndim() == 4)
+  if (common::ContainsOnlyStorage(inputs, kCSRStorage)) {
```
Comment (eric-haibin-lin) [inline src/operator/nn/concat.cc]: common::ContainsOnlyStorage(inputs, kCSRStorage) && output.stype == kCSR

**Comment by ZiyueHuang** [comment]
good to merge? @eric-haibin-lin

File: src/operator/nn/concat.cc (hunk 13 above)
Comment (pengzhao-intel) [reply src/operator/nn/concat.cc]: Yes, MKL-DNN `concat` only supports the dim >0 now. Will take a look at the current flow.

**Pull request merged**